File: leetcode/105.从前序与中序遍历序列构造二叉树.go
/*
 * @lc app=leetcode.cn id=105 lang=golang
 *
 * [105] 从前序与中序遍历序列构造二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 //迭代法
func buildTree(preorder []int, inorder []int) *TreeNode {
	if preorder==nil||len(preorder)==0{
		return nil
	}
	var length int = len(preorder)
	var inIdx = 0
	var root *TreeNode = &TreeNode{
		Val:preorder[0],
		Left:nil,
		Right:nil,
	}
	//用slice当作栈，存放前序遍历的节点
	var stack []*TreeNode = make([]*TreeNode, 0)
	//push操作，将节点压入栈
	stack = append(stack,root)
	for i := 1; i < length; i++ {
		//peek操作，访问栈顶元素
		var node *TreeNode = stack[len(stack)-1]
		if node.Val != inorder[inIdx]{
           node.Left = &TreeNode{
			Val:preorder[i],
			Left:nil,
			Right:nil,
		   }
		   //push操作，将节点压入栈
		   stack = append(stack,node.Left)
		}else{
            for len(stack)!=0&&
			stack[len(stack)-1].Val == inorder[inIdx]{
              node = stack[len(stack)-1]
			  //pop操作，将栈顶元素弹出
			  stack = stack[0:len(stack)-1]
			  inIdx +=1
			}
            node.Right = &TreeNode{
				Val:preorder[i],
				Left:nil,
				Right:nil,
			   }
			   //节点入栈
			stack = append(stack,node.Right)
		}
	}
	return root
}

// @lc code=end

File: leetcode/105.从前序与中序遍历序列构造二叉树.java
/*
 * @lc app=leetcode.cn id=105 lang=java
 *
 * [105] 从前序与中序遍历序列构造二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val; 
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int[] preorder;
    int preIdx;
    Map<Integer,Integer> map = new HashMap<>();
    public TreeNode buildTree(int[] preorder, int[] inorder) {
     this.preorder = preorder;
     this.preIdx = 0;
     for(int i=0,len=inorder.length;i<len;++i){
         map.put(inorder[i],i);
     }
     return build(0,inorder.length);
    }
    //递归函数
    private TreeNode build(int left,int right){
        if (left >= right){
		return null;
	}
	int root_val = preorder[preIdx];
	TreeNode root = new TreeNode(root_val);
	int mid = map.get(root_val);
	preIdx++;
	root.left = build(left,mid);
	root.right = build(mid+1,right);
	return root;
    }
}
// @lc code=end

